test/balancer_test.go

```diff
@@ -981,3 +981,266 @@ func (s) TestSubConnShutdown(t *testing.T) {
+		grpc.WithDefaultServiceConfig(svcCfg),
+	}
+	cc, err := grpc.NewClient(backend.Address, opts...)
+	if err != nil {
+		t.Fatalf("grpc.NewClient(%q) failed: %v", backend.Address, err)
+
+	}
+	defer cc.Close()
+
+	cc.Connect()
+
+	var sc balancer.SubConn
+	select {
+	case sc = <-scChan:
+	case <-ctx.Done():
+		t.Fatal("Context timed out waiting for SubConn creation")
+	}
+
+	// Wait for the SubConn to enter READY.
+	select {
+	case <-readyUpdateReceivedCh:
+	case <-ctx.Done():
+		t.Fatalf("Context timed out waiting for SubConn to enter READY")
+	}
+
+	healthChan1 := make(chan balancer.SubConnState, 1)
+	healthChan2 := make(chan balancer.SubConnState, 1)
+
+	sc.RegisterHealthListener(func(scs balancer.SubConnState) {
+		healthChan1 <- scs
+	})
+	sc.RegisterHealthListener(func(scs balancer.SubConnState) {
+		healthChan2 <- scs
+	})
+	close(readyUpdateResumeCh)
+
+	select {
+	case scs := <-healthChan2:
+		if scs.ConnectivityState != connectivity.Ready {
+			t.Fatalf("Received health update = %v, want = %v", scs.ConnectivityState, connectivity.Ready)
+		}
+	case <-ctx.Done():
+		t.Fatalf("Context timed out waiting for health update")
+	}
+
+	// No updates should be received on the first listener.
+	select {
+	case scs := <-healthChan1:
+		t.Fatalf("Received unexpected health update on first listener: %v", scs)
+	case <-time.After(defaultTestShortTimeout):
+	}
+}
+
+// Test calls RegisterHealthListener on a SubConn with a nil listener and
+// verifies that it doesn't panic.
+func (s) TestSubConn_RegisterHealthListener_NilListener(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	scChan := make(chan balancer.SubConn, 1)
+	bf := stub.BalancerFuncs{
+		Init: func(bd *stub.BalancerData) {
+			cc := bd.ClientConn
+			ccw := &subConnStoringCCWrapper{
+				ClientConn: cc,
+				scChan:     scChan,
+			}
+			bd.Data = balancer.Get(pickfirst.Name).Build(ccw, bd.BuildOptions)
+		},
+		Close: func(bd *stub.BalancerData) {
+			bd.Data.(balancer.Balancer).Close()
+		},
+		UpdateClientConnState: func(bd *stub.BalancerData, ccs balancer.ClientConnState) error {
+			return bd.Data.(balancer.Balancer).UpdateClientConnState(ccs)
+		},
+	}
+
+	stub.Register(t.Name(), bf)
+	svcCfg := fmt.Sprintf(`{ "loadBalancingConfig": [{%q: {}}] }`, t.Name())
+	backend := stubserver.StartTestService(t, nil)
+	defer backend.Stop()
+	opts := []grpc.DialOption{
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithDefaultServiceConfig(svcCfg),
+	}
+	cc, err := grpc.NewClient(backend.Address, opts...)
+	if err != nil {
+		t.Fatalf("grpc.NewClient(%q) failed: %v", backend.Address, err)
+
+	}
+	defer cc.Close()
+
+	cc.Connect()
+
+	var sc balancer.SubConn
+	select {
+	case sc = <-scChan:
+	case <-ctx.Done():
+		t.Fatal("Context timed out waiting for SubConn creation")
+	}
+
+	testutils.AwaitState(ctx, t, cc, connectivity.Ready)
+	sc.RegisterHealthListener(nil)
```

Comment:
How about a test case to ensure that a previously registered non-nil listener (that is yet to be invoked) is dropped when a `nil` listener is registered?

Reply:
Updated the test case to verify that the nil listener causes the original listener to be dropped.